server/webapp/WEB-INF/rails.new/app/controllers/admin/materials/pluggable_scm_controller.rb

```diff
@@ -0,0 +1,212 @@
+
+      create_failure_handler = proc do |result, all_errors|
+        @errors = flatten_all_errors(all_errors)
+        @meta_data_store = meta_data_store
+        render :template => '/admin/materials/pluggable_scm/edit', :status => result.httpCode(), :layout => false
+      end
+
+      save_popup(params[:config_md5], get_update_command, create_failure_handler, { :controller => '/admin/materials', :stage_parent => "pipelines", :current_tab => params[:current_tab] }) do
+        assert_load :pipeline, @node.pipelineConfigByName(CaseInsensitiveString.new(params[:pipeline_name]))
+        assert_load :material, @subject
+      end
+    end
+
+    def check_connection
+      scm = com.thoughtworks.go.domain.scm.SCM.new
+      scm.setPluginConfiguration(PluginConfiguration.new(params[:plugin_id], nil))
+      scm.setConfigAttributes(params[:material])
+
+      result = pluggable_scm_service.checkConnection(scm)
+      render json: { status: result.isSuccessful() ? 'success' : 'failure', messages: result.getMessages() }
+    end
+
+    private
+
+    def get_choose_existing_command
+      Class.new(::ConfigUpdate::SaveAsPipelineAdmin) do
+        include ::ConfigUpdate::LoadConfig
+
+        def initialize params, user, security_service, material
+          super(params, user, security_service)
+          @material = material
+        end
+
+        def node(cruise_config)
+          cruise_config
+        end
+
+        def subject(cruise_config)
+          cruise_config.pipelineConfigByName(CaseInsensitiveString.new(params[:pipeline_name])).materialConfigs().last()
+        end
+
+        def update(cruise_config)
+          scm = cruise_config.getSCMs().find(params[:material][:scmId])
+          @material.setSCMConfig(scm)
+
+          @material.setConfigAttributes(params[:material])
+
+          cruise_config.pipelineConfigByName(CaseInsensitiveString.new(params[:pipeline_name])).addMaterialConfig(@material)
+        end
+      end.new(params, current_user.getUsername(), security_service, @material)
+    end
+
+    def get_create_command
+      Class.new(::ConfigUpdate::SaveAsPipelineAdmin) do
+        include ::ConfigUpdate::LoadConfig
+
+        def initialize params, user, security_service, material, pluggable_scm_service
+          super(params, user, security_service)
+          @material = material
+          @pluggable_scm_service = pluggable_scm_service
+        end
+
+        def node(cruise_config)
+          cruise_config
+        end
+
+        def subject(cruise_config)
+          cruise_config.pipelineConfigByName(CaseInsensitiveString.new(params[:pipeline_name])).materialConfigs().last()
+        end
+
+        def update(cruise_config)
+          scm = @material.getSCMConfig()
+          scm.setConfigAttributes(params[:material])
+          scm.ensureIdExists
+          params[:material][PluggableSCMMaterialConfig::SCM_ID] = scm.getSCMId
+
+          @pluggable_scm_service.validate(scm)
```

Comment:
Why is this continuing even after validate might have failed? Is that the pattern in this whole save_popup area? I guess it will fail after, while validating everything. Sigh.